[BEAM-13314]Revise recommendations to manage Python pipeline dependencies. (#16938)
apache/beam: website/www/site/content/en/documentation/runtime/environments.md

By default, no licenses/notices are added to the docker images.

#### Modifying an existing container image to make it compatible with Apache Beam Runners {#modify-existing-base-image}

> Beam offers a way to provide your own custom container image. The easiest way to build a new custom image that is compatible with Apache Beam Runners is to use a [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) process. This copies over the necessary artifacts from a default Apache Beam base image to build your custom container image.

Reviewer: Is this multi-stage build?

Author: Yes. Using multistaging build process to copy required artifacts from Apache Beam's base image to the provided custom image

1. Copy necessary artifacts from Apache Beam base image to your image.
  ```
  # This can be any container image,
 FROM python:3.7-bullseye

 # Install SDK. (needed for Python SDK)
 RUN pip install --no-cache-dir apache-beam[gcp]==2.35.0

 # Copy files from official SDK image, including script/dependencies.

>  COPY --from=apache/beam_python3.7_sdk:2.35.0 /opt/apache/beam /opt/apache/beam

Reviewer: I wonder if this is sufficient, IIRC /opt/apache/beam this only contains the boot program? all the base_image_requirements are in site_packages?

Author: they are in dist_packages

Reviewer 2: the requirements will be installed at runtime , when we pip-install the staged apache beam sdk here https://github.com/apache/beam/blob/d46bd07501e01c7d35d3857e296bc8b75564a8f0/sdks/python/container/piputil.go#L164

 # Perform any additional customizations if desired

 # Set the entrypoint to Apache Beam SDK launcher.
 ENTRYPOINT ["/opt/apache/beam/boot"]

  ```
>**NOTE**: This example assumes necessary dependencies (in this case, Python 3.7 and pip) have been installed on the existing base image. Installing the Apache Beam SDK into the image will ensure that the image has the necessary SDK dependencies and reduce the worker startup time.

> >The version specified in the `RUN` instruction must match the version used to launch the pipeline.<br>

Reviewer 2: I'd expand this to smth like:

The Apache Beam version and the Python version specified in the custom Dockerfile must match the versions  used to launch the pipeline.

Reviewer 2: I see you mention this below

> >**Users need to make sure that whatever base image they use has the same Python/Java interpreter version that they used to run the pipeline**.

Reviewer 3: Consider: "Make sure that the Python or Java runtime version specified in the base image is the same as the version used to run the pipeline."

By default, no licenses/notices are added to the docker images.

> #### Build an existing container image to make it compatible with Apache Beam Runners {#modify-existing-base-image}

Reviewer 2: @emilymye could you PTAL at this section?

> #### Build an existing container image to make it compatible with Apache Beam Runners {#modify-existing-base-image}

Reviewer 4: ```suggestion
#### Build a container image based on an existing image compatible with Apache Beam Runners  {#modify-existing-base-image}
```

> Beam offers a way to take a Beam container image and customize it. But if you have an existing base image to be compatible with Apache Beam Runners, use a [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) process to copy over the necessary artifacts from a default Apache Beam base image and provide your custom container image.

Reviewer 4: ```suggestion
Beam offers a way to provide your own custom Beam container image. The easiest way to build a new custom image that is compatible with Apache Beam Runners is to use a [multi-stage build](https://docs.docker.com/develop/develop-images/multistage-build/) process. This copies over the necessary artifacts from a default Apache Beam base image to build your custom container image.
```

1. Copy necessary artifacts from Apache Beam base image to your image.
  ```
  # This can be any container image,
 FROM python:3.8-slim

 # Install SDK. (needed for Python SDK)

>  RUN pip install --no-cache-dir apache-beam[gcp]==2.25.0

Reviewer 4: Can make the version here a build arg https://docs.docker.com/engine/reference/builder/#arg

e.g. 

```
FROM python:3.8-slim
ARG beam_ver=2.25.0
...

RUN pip install --no-cache-dir apache-beam[gcp]==$beam_ver
COPY --from=apache/beam_python3.7_sdk:$beam_ver /opt/apache/beam /opt/apache/beam
```

Reviewer 2: I think we tried this before and it didn't work.

Author: I built a Dockerfile with `ARG BEAM_VERSION=2.31.0` and `RUN pip install --no-cache-dir  apache-beam[gcp]==$BEAM_VERSION`. This seems to be working. I verified the /opt/apache/beam/boot path. 

Can you recall what was the error?

Author: Okay. If I use ARG and in the docker container, I tried 'import apache_beam'. I got this error.

```
>>> import apache_beam
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/__init__.py", line 93, in <module>
    from apache_beam import io
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/io/__init__.py", line 21, in <module>
    from apache_beam.io.avroio import *
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/io/avroio.py", line 53, in <module>
    from apache_beam.io import filebasedsink
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/io/filebasedsink.py", line 29, in <module>
    from apache_beam.io import iobase
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/io/iobase.py", line 54, in <module>
    from apache_beam.transforms import Impulse
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/transforms/__init__.py", line 24, in <module>
    from apache_beam.transforms.external import *
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/transforms/external.py", line 44, in <module>
    from apache_beam.runners import pipeline_context
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/runners/__init__.py", line 29, in <module>
    from apache_beam.runners.dataflow.dataflow_runner import DataflowRunner
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/runners/dataflow/__init__.py", line 24, in <module>
    from apache_beam.runners.dataflow.dataflow_runner import DataflowRunner
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/runners/dataflow/dataflow_runner.py", line 43, in <module>
    from apache_beam.internal.gcp import json_value
  File "/usr/local/lib/python3.6/dist-packages/apache_beam/internal/gcp/json_value.py", line 27, in <module>
    from apitools.base.py import extra_types
  File "/usr/local/lib/python3.6/dist-packages/apitools/base/py/__init__.py", line 21, in <module>
    from apitools.base.py.base_api import *
  File "/usr/local/lib/python3.6/dist-packages/apitools/base/py/base_api.py", line 35, in <module>
    from apitools.base.py import http_wrapper
  File "/usr/local/lib/python3.6/dist-packages/apitools/base/py/http_wrapper.py", line 29, in <module>
    import httplib2
  File "/usr/local/lib/python3.6/dist-packages/httplib2/__init__.py", line 52, in <module>
    from . import auth
  File "/usr/local/lib/python3.6/dist-packages/httplib2/auth.py", line 20, in <module>
    auth_param_name = token.copy().setName("auth-param-name").addParseAction(pp.downcaseTokens)
AttributeError: module 'pyparsing' has no attribute 'downcaseTokens'
```

Reviewer 2: I thought using Dockerfile params (ARG) in the FROM statement wasn't possible. That's what I was referring to.

Author: Yes, expansion of ARGS in --from is not supported in Docker. Hence moving forward with the current instructions

 # Copy files from official SDK image, including script/dependencies.
 COPY --from=apache/beam_python3.7_sdk:2.25.0 /opt/apache/beam /opt/apache/beam

>  # Perform any addtional customizations if desired

Reviewer 4: ```suggestion
 # Perform any additional customizations if desired
```
